File: MessagesKit/DataReference.m
//
//  DataReference.m
//  MessagesKit
//
//  Created by Kevin Wooten on 4/25/16.
//  Copyright © 2016 reTXT Labs LLC. All rights reserved.
//

#import "DataReference.h"


@implementation NSInputStream (DataReference)

-(NSUInteger) availableBytes
{
  return self.hasBytesAvailable ? 1 : 0;
}

-(BOOL) readBytesOfMaxLength:(NSUInteger)maxLength intoBuffer:(UInt8 *)buffer bytesRead:(NSUInteger *)bytesRead error:(NSError **)error
{
  NSInteger res = [self read:buffer maxLength:maxLength];
  if (res < 0) {
    if (error) {
      *error = self.streamError ? self.streamError : [NSError errorWithDomain:@"NSStreamErrorDomain"
                                                                         code:0
                                                                     userInfo:@{NSLocalizedDescriptionKey:@"Read failed"}];
    }
    return NO;
  }
  *bytesRead = res;
  return YES;
}

@end



@implementation NSOutputStream (DataReference)

-(BOOL) writeBytesFromBuffer:(const UInt8 *)buffer length:(NSUInteger)length error:(NSError * _Nullable __autoreleasing *)error
{
  NSUInteger written = [self write:buffer maxLength:length];
  if (written != length) {
    if (error) {
      *error = self.streamError ? self.streamError : [NSError errorWithDomain:@"NSStreamErrorDomain"
                                                                         code:0
                                                                     userInfo:@{NSLocalizedDescriptionKey:@"Write failed"}];
    }
    return NO;
  }
  return YES;
}

@end

File: MessagesKit/DataReference.swift
//
//  DataReference.swift
//  MessagesKit
//
//  Created by Kevin Wooten on 4/21/16.
//  Copyright © 2016 reTXT Labs LLC. All rights reserved.
//

import Foundation
import FMDB
import Darwin
import MobileCoreServices
 

public typealias DataReferenceFilter = (DataInputStream, DataOutputStream) throws -> Void

extension DataReference {
  
  public func temporaryDuplicate(MIMEType: String? = nil, filteredBy filter: DataReferenceFilter? = nil) throws -> DataReference {
    
    guard let filter = filter else {
      return try self.__temporaryDuplicateFilteredBy(nil, withMIMEType: MIMEType)
    }
    
    return try self.__temporaryDuplicateFilteredBy({ ins, outs, error in
      do {
        try filter(ins, outs)
        return true
      }
      catch let caughtError {
        error.memory = caughtError as NSError
        return false
      }
    }, withMIMEType: MIMEType)
  }

  public func readAllData() throws -> NSData {
    return try DataReferences.readAllDataFromReference(self)
  }
  
  public func saveToTemporaryURL() throws -> NSURL {
    return try DataReferences.saveDataReferenceToTemporaryURL(self)
  }
  
}
